### PR TITLE
[build][Zig] Make more flags from `src/config.h` configurable

### PR DESCRIPTION
**Pull request by sagehane**
This PR does not depend on #4406.

Currently, only flags containing the word `SUPPORT` is override-able via `-Dconfig`. This was the case as there was no code to check for what value they were set to, and all `SUPPORT` flags have a value of 1. This is also not ideal because it'll cause unexpected behaviour if the convention of the current `src/config.h` changes.

With this PR, everything that isn't defined between `#ifdef` or `#ifndef` and `#endif` can be passed or configured. I didn't bother evaluating flags between those as they already seem to be meant to be configured and truly evaluating them would be very annoying.

List of flags that are now passed by default when passing `-Dconfig`:
<details><pre>
-DSUPPORT_MODULE_RSHAPES=1
-DSUPPORT_MODULE_RTEXTURES=1
-DSUPPORT_MODULE_RTEXT=1
-DSUPPORT_MODULE_RMODELS=1
-DSUPPORT_MODULE_RAUDIO=1
-DSUPPORT_CAMERA_SYSTEM=1
-DSUPPORT_GESTURES_SYSTEM=1
-DSUPPORT_RPRAND_GENERATOR=1
-DSUPPORT_MOUSE_GESTURES=1
-DSUPPORT_SSH_KEYBOARD_RPI=1
-DSUPPORT_WINMM_HIGHRES_TIMER=1
-DSUPPORT_PARTIALBUSY_WAIT_LOOP=1
-DSUPPORT_SCREEN_CAPTURE=1
-DSUPPORT_GIF_RECORDING=1
-DSUPPORT_COMPRESSION_API=1
-DSUPPORT_AUTOMATION_EVENTS=1
-DMAX_FILEPATH_CAPACITY=8192
-DMAX_FILEPATH_LENGTH=4096
-DMAX_KEYBOARD_KEYS=512
-DMAX_MOUSE_BUTTONS=8
-DMAX_GAMEPADS=4
-DMAX_GAMEPAD_AXIS=8
-DMAX_GAMEPAD_BUTTONS=32
-DMAX_GAMEPAD_VIBRATION_TIME=2.0f
-DMAX_TOUCH_POINTS=8
-DMAX_KEY_PRESSED_QUEUE=16
-DMAX_CHAR_PRESSED_QUEUE=16
-DMAX_DECOMPRESSION_SIZE=64
-DMAX_AUTOMATION_EVENTS=16384
-DRL_DEFAULT_BATCH_BUFFERS=1
-DRL_DEFAULT_BATCH_DRAWCALLS=256
-DRL_DEFAULT_BATCH_MAX_TEXTURE_UNITS=4
-DRL_MAX_MATRIX_STACK_SIZE=32
-DRL_MAX_SHADER_LOCATIONS=32
-DRL_CULL_DISTANCE_NEAR=0.01
-DRL_CULL_DISTANCE_FAR=1000.0
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_POSITION=0
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_TEXCOORD=1
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_NORMAL=2
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_COLOR=3
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_TANGENT=4
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_TEXCOORD2=5
-DRL_DEFAULT_SHADER_ATTRIB_LOCATION_INDICES=6
-DRL_SUPPORT_MESH_GPU_SKINNING
-DRL_DEFAULT_SHADER_ATTRIB_NAME_POSITION="vertexPosition"
-DRL_DEFAULT_SHADER_ATTRIB_NAME_TEXCOORD="vertexTexCoord"
-DRL_DEFAULT_SHADER_ATTRIB_NAME_NORMAL="vertexNormal"
-DRL_DEFAULT_SHADER_ATTRIB_NAME_COLOR="vertexColor"
-DRL_DEFAULT_SHADER_ATTRIB_NAME_TANGENT="vertexTangent"
-DRL_DEFAULT_SHADER_ATTRIB_NAME_TEXCOORD2="vertexTexCoord2"
-DRL_DEFAULT_SHADER_UNIFORM_NAME_MVP="mvp"
-DRL_DEFAULT_SHADER_UNIFORM_NAME_VIEW="matView"
-DRL_DEFAULT_SHADER_UNIFORM_NAME_PROJECTION="matProjection"
-DRL_DEFAULT_SHADER_UNIFORM_NAME_MODEL="matModel"
-DRL_DEFAULT_SHADER_UNIFORM_NAME_NORMAL="matNormal"
-DRL_DEFAULT_SHADER_UNIFORM_NAME_COLOR="colDiffuse"
-DRL_DEFAULT_SHADER_SAMPLER2D_NAME_TEXTURE0="texture0"
-DRL_DEFAULT_SHADER_SAMPLER2D_NAME_TEXTURE1="texture1"
-DRL_DEFAULT_SHADER_SAMPLER2D_NAME_TEXTURE2="texture2"
-DSUPPORT_QUADS_DRAW_MODE=1
-DSPLINE_SEGMENT_DIVISIONS=24
-DSUPPORT_FILEFORMAT_PNG=1
-DSUPPORT_FILEFORMAT_GIF=1
-DSUPPORT_FILEFORMAT_QOI=1
-DSUPPORT_FILEFORMAT_DDS=1
-DSUPPORT_IMAGE_EXPORT=1
-DSUPPORT_IMAGE_GENERATION=1
-DSUPPORT_IMAGE_MANIPULATION=1
-DSUPPORT_DEFAULT_FONT=1
-DSUPPORT_FILEFORMAT_TTF=1
-DSUPPORT_FILEFORMAT_FNT=1
-DSUPPORT_TEXT_MANIPULATION=1
-DSUPPORT_FONT_ATLAS_WHITE_REC=1
-DMAX_TEXT_BUFFER_LENGTH=1024
-DMAX_TEXTSPLIT_COUNT=128
-DSUPPORT_FILEFORMAT_OBJ=1
-DSUPPORT_FILEFORMAT_MTL=1
-DSUPPORT_FILEFORMAT_IQM=1
-DSUPPORT_FILEFORMAT_GLTF=1
-DSUPPORT_FILEFORMAT_VOX=1
-DSUPPORT_FILEFORMAT_M3D=1
-DSUPPORT_MESH_GENERATION=1
-DMAX_MATERIAL_MAPS=12
-DSUPPORT_FILEFORMAT_WAV=1
-DSUPPORT_FILEFORMAT_OGG=1
-DSUPPORT_FILEFORMAT_MP3=1
-DSUPPORT_FILEFORMAT_QOA=1
-DSUPPORT_FILEFORMAT_XM=1
-DSUPPORT_FILEFORMAT_MOD=1
-DAUDIO_DEVICE_FORMAT=ma_format_f32
-DAUDIO_DEVICE_CHANNELS=2
-DAUDIO_DEVICE_SAMPLE_RATE=0
-DMAX_AUDIO_BUFFER_POOL_CHANNELS=16
-DSUPPORT_STANDARD_FILEIO=1
-DSUPPORT_TRACELOG=1
-DMAX_TRACELOG_MSG_LENGTH=256
</pre></details>

### CODE REVIEW

File: build.zig [hunk 1]
```diff
@@ -114,8 +126,11 @@ fn compileRaylib(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.
         // Note: Currently using a suboptimal `O(m*n)` time algorithm where:
         // `m` corresponds roughly to the number of lines in `src/config.h`
         // `n` corresponds to the number of user-specified flags
-        outer: for (config_h_flags) |flag| {
+        outer: for (config_h_flags) |flag_val| {
+            const flag = flag_val[0];
+
             // If a user already specified the flag, skip it
+            config_iter.reset();
```
Comment (sagehane) [inline build.zig:133]: This line was supposed to be here since #4398, but I somehow forgot to include it. Without it, one can pass custom flags but won't be able to properly override existing ones: Both the default and overriding flags will end up being passed.

**Comment by sagehane** [comment]
Side note: I wonder if using `std.Build.Step.Compile.defineCMacro` is more or less idiomatic than passing these flags the way it is. My guess that it's more idiomatic, so I'll switch to that approach.

**Comment by sagehane** [comment]
Should be ready to merge; just needs reviews.

File: build.zig [hunk 2]
```diff
@@ -237,9 +258,9 @@ fn compileRaylib(b: *std.Build, target: std.Build.ResolvedTarget, optimize: std.
                 raylib.linkSystemLibrary("dl");
                 raylib.addIncludePath(.{ .cwd_relative = "/usr/include/libdrm" });
 
-                raylib.defineCMacro("PLATFORM_DRM", null);
-                raylib.defineCMacro("EGL_NO_X11", null);
-                raylib.defineCMacro("DEFAULT_BATCH_BUFFER_ELEMENT", "2048");
+                raylib.root_module.addCMacro("PLATFORM_DRM", "1");
+                raylib.root_module.addCMacro("EGL_NO_X11", "1");
+                raylib.root_module.addCMacro("DEFAULT_BATCH_BUFFER_ELEMENT", "2048");
```
Comment (sagehane) [inline build.zig:263]: I wonder what this `DEFAULT_BATCH_BUFFER_ELEMENT` is. `src/config.h` mentions `RL_DEFAULT_BATCH_BUFFER_ELEMENTS` [here](https://github.com/raysan5/raylib/blob/5.0/src/config.h#L102) but it's a slightly different name, off by default, and not the same value.

@raysan5, do you have any ideas? Is this some old deprecated name for the setting? And does it really need to be defined when building for `-Dplatform=drm`? I don't see it mentioned by any other build script either.

Comment (raysan5) [reply build.zig:263]: I think it's a deprecated name, it shouldn't be needed... I think... rlgl has a definition for it dependant on OpenGL version

Comment (sagehane) [reply build.zig:263]: Yeah... I'll remove it. At a glance, it's the only suspicious config aside from the undocumented one at:
https://github.com/raysan5/raylib/blob/6610ab0dc73f87dc836526e0565e30d77faf0623/build.zig#L125
OpenGL deprecation... maybe that's some macOS thing.

**Comment by raysan5** [comment]
@sagehane What is the state of this PR? Can it be merged?

File: build.zig (hunk 2 above)
Comment (Not-Nik) [reply build.zig:263]: I think that just silences build warnings when calling OpenGL functions, which obviously is macOS specific, so it should be fine to only pass there.

**Comment by sagehane** [comment]
> @sagehane What is the state of this PR? Can it be merged?

It's mostly okay but has a minor conflict because I extracted a portion of this patch to a different PR. I haven't had time to fix it and I don't know if I'll have much time for this until the end of the month.

That said, this is an enhancement that could use testing and not a bug fix of some immediate issue so it could wait until you release a new version like you said you're planning to. Feel free to close it if you don't want noise in the PR tracker; I can just reopen it after you tag a new release.

**Comment by sagehane** [comment]
Okay, I finally had the time to update this PR.

Here's the flags passed now:
<details><pre>
@as(*const [24:0]u8, "SUPPORT_MODULE_RSHAPES=1")
@as(*const [26:0]u8, "SUPPORT_MODULE_RTEXTURES=1")
@as(*const [22:0]u8, "SUPPORT_MODULE_RTEXT=1")
@as(*const [24:0]u8, "SUPPORT_MODULE_RMODELS=1")
@as(*const [23:0]u8, "SUPPORT_MODULE_RAUDIO=1")
@as(*const [23:0]u8, "SUPPORT_CAMERA_SYSTEM=1")
@as(*const [25:0]u8, "SUPPORT_GESTURES_SYSTEM=1")
@as(*const [26:0]u8, "SUPPORT_RPRAND_GENERATOR=1")
@as(*const [24:0]u8, "SUPPORT_MOUSE_GESTURES=1")
@as(*const [26:0]u8, "SUPPORT_SSH_KEYBOARD_RPI=1")
@as(*const [29:0]u8, "SUPPORT_WINMM_HIGHRES_TIMER=1")
@as(*const [31:0]u8, "SUPPORT_PARTIALBUSY_WAIT_LOOP=1")
@as(*const [24:0]u8, "SUPPORT_SCREEN_CAPTURE=1")
@as(*const [23:0]u8, "SUPPORT_GIF_RECORDING=1")
@as(*const [25:0]u8, "SUPPORT_COMPRESSION_API=1")
@as(*const [27:0]u8, "SUPPORT_AUTOMATION_EVENTS=1")
@as(*const [26:0]u8, "MAX_FILEPATH_CAPACITY=8192")
@as(*const [24:0]u8, "MAX_FILEPATH_LENGTH=4096")
@as(*const [21:0]u8, "MAX_KEYBOARD_KEYS=512")
@as(*const [19:0]u8, "MAX_MOUSE_BUTTONS=8")
@as(*const [14:0]u8, "MAX_GAMEPADS=4")
@as(*const [18:0]u8, "MAX_GAMEPAD_AXIS=8")
@as(*const [22:0]u8, "MAX_GAMEPAD_BUTTONS=32")
@as(*const [31:0]u8, "MAX_GAMEPAD_VIBRATION_TIME=2.0f")
@as(*const [18:0]u8, "MAX_TOUCH_POINTS=8")
@as(*const [24:0]u8, "MAX_KEY_PRESSED_QUEUE=16")
@as(*const [25:0]u8, "MAX_CHAR_PRESSED_QUEUE=16")
@as(*const [25:0]u8, "MAX_DECOMPRESSION_SIZE=64")
@as(*const [27:0]u8, "MAX_AUTOMATION_EVENTS=16384")
@as(*const [30:0]u8, "RL_SUPPORT_MESH_GPU_SKINNING=1")
@as(*const [26:0]u8, "RL_DEFAULT_BATCH_BUFFERS=1")
@as(*const [30:0]u8, "RL_DEFAULT_BATCH_DRAWCALLS=256")
@as(*const [36:0]u8, "RL_DEFAULT_BATCH_MAX_TEXTURE_UNITS=4")
@as(*const [27:0]u8, "RL_MAX_MATRIX_STACK_SIZE=32")
@as(*const [26:0]u8, "RL_MAX_SHADER_LOCATIONS=32")
@as(*const [26:0]u8, "RL_CULL_DISTANCE_NEAR=0.01")
@as(*const [27:0]u8, "RL_CULL_DISTANCE_FAR=1000.0")
@as(*const [44:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_POSITION=0")
@as(*const [44:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_TEXCOORD=1")
@as(*const [42:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_NORMAL=2")
@as(*const [41:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_COLOR=3")
@as(*const [43:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_TANGENT=4")
@as(*const [45:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_TEXCOORD2=5")
@as(*const [43:0]u8, "RL_DEFAULT_SHADER_ATTRIB_LOCATION_INDICES=6")
@as(*const [55:0]u8, "RL_DEFAULT_SHADER_ATTRIB_NAME_POSITION=\"vertexPosition\"")
@as(*const [55:0]u8, "RL_DEFAULT_SHADER_ATTRIB_NAME_TEXCOORD=\"vertexTexCoord\"")
@as(*const [51:0]u8, "RL_DEFAULT_SHADER_ATTRIB_NAME_NORMAL=\"vertexNormal\"")
@as(*const [49:0]u8, "RL_DEFAULT_SHADER_ATTRIB_NAME_COLOR=\"vertexColor\"")
@as(*const [53:0]u8, "RL_DEFAULT_SHADER_ATTRIB_NAME_TANGENT=\"vertexTangent\"")
@as(*const [57:0]u8, "RL_DEFAULT_SHADER_ATTRIB_NAME_TEXCOORD2=\"vertexTexCoord2\"")
@as(*const [40:0]u8, "RL_DEFAULT_SHADER_UNIFORM_NAME_MVP=\"mvp\"")
@as(*const [45:0]u8, "RL_DEFAULT_SHADER_UNIFORM_NAME_VIEW=\"matView\"")
@as(*const [57:0]u8, "RL_DEFAULT_SHADER_UNIFORM_NAME_PROJECTION=\"matProjection\"")
@as(*const [47:0]u8, "RL_DEFAULT_SHADER_UNIFORM_NAME_MODEL=\"matModel\"")
@as(*const [49:0]u8, "RL_DEFAULT_SHADER_UNIFORM_NAME_NORMAL=\"matNormal\"")
@as(*const [49:0]u8, "RL_DEFAULT_SHADER_UNIFORM_NAME_COLOR=\"colDiffuse\"")
@as(*const [52:0]u8, "RL_DEFAULT_SHADER_SAMPLER2D_NAME_TEXTURE0=\"texture0\"")
@as(*const [52:0]u8, "RL_DEFAULT_SHADER_SAMPLER2D_NAME_TEXTURE1=\"texture1\"")
@as(*const [52:0]u8, "RL_DEFAULT_SHADER_SAMPLER2D_NAME_TEXTURE2=\"texture2\"")
@as(*const [25:0]u8, "SUPPORT_QUADS_DRAW_MODE=1")
@as(*const [27:0]u8, "SPLINE_SEGMENT_DIVISIONS=24")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_PNG=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_GIF=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_QOI=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_DDS=1")
@as(*const [22:0]u8, "SUPPORT_IMAGE_EXPORT=1")
@as(*const [26:0]u8, "SUPPORT_IMAGE_GENERATION=1")
@as(*const [28:0]u8, "SUPPORT_IMAGE_MANIPULATION=1")
@as(*const [22:0]u8, "SUPPORT_DEFAULT_FONT=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_TTF=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_FNT=1")
@as(*const [27:0]u8, "SUPPORT_TEXT_MANIPULATION=1")
@as(*const [30:0]u8, "SUPPORT_FONT_ATLAS_WHITE_REC=1")
@as(*const [27:0]u8, "MAX_TEXT_BUFFER_LENGTH=1024")
@as(*const [23:0]u8, "MAX_TEXTSPLIT_COUNT=128")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_OBJ=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_MTL=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_IQM=1")
@as(*const [25:0]u8, "SUPPORT_FILEFORMAT_GLTF=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_VOX=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_M3D=1")
@as(*const [25:0]u8, "SUPPORT_MESH_GENERATION=1")
@as(*const [20:0]u8, "MAX_MATERIAL_MAPS=12")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_WAV=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_OGG=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_MP3=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_QOA=1")
@as(*const [23:0]u8, "SUPPORT_FILEFORMAT_XM=1")
@as(*const [24:0]u8, "SUPPORT_FILEFORMAT_MOD=1")
@as(*const [33:0]u8, "AUDIO_DEVICE_FORMAT=ma_format_f32")
@as(*const [23:0]u8, "AUDIO_DEVICE_CHANNELS=2")
@as(*const [26:0]u8, "AUDIO_DEVICE_SAMPLE_RATE=0")
@as(*const [33:0]u8, "MAX_AUDIO_BUFFER_POOL_CHANNELS=16")
@as(*const [25:0]u8, "SUPPORT_STANDARD_FILEIO=1")
@as(*const [18:0]u8, "SUPPORT_TRACELOG=1")
@as(*const [27:0]u8, "MAX_TRACELOG_MSG_LENGTH=256")
@as(*const [25:0]u8, "SUPPORT_CLIPBOARD_IMAGE=1")
</pre></details>

It's worth noting that these newly introduced lines are hard to parse properly due to them being defined inside a conditional block.
https://github.com/raysan5/raylib/blob/00396e34369a1696180849216c3e5aca264c70e4/src/config.h#L279-L301

The project currently successfully compiles when passing `zig build -Dconfig="-DSTBI_REQUIRED -DSUPPORT_FILEFORMAT_BMP"`.

I think it can still be treated as a draft until a new release comes out.

**Comment by raysan5** [comment]
@sagehane It seems this PR conflicts with https://github.com/raysan5/raylib/pull/4620, could we close this PR and re-submit it again onece #4620 gets merged?

**Comment by sagehane** [comment]
Yes, that sounds fine by me. In fact, I think the current approach for passing custom macro values is quite ugly and can be better addressed once/if https://github.com/raysan5/raylib/issues/4411 is resolved.
